### PR TITLE
add WeatherServiceTest

### PR DESCRIPTION
**Pull request by alesyalis**

### CODE REVIEW

File: Weather.СonsoleApp/Weather.Tests/Service/WeatherServiceTest.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+            _weatherRepositoryMock = new Mock<IWeatherRepository>();
+            _validatorMock = new Mock<IValidator>();    
+
+            _weatherService = new WeatherService(
+                _weatherRepositoryMock.Object,
+                _validatorMock.Object);
+        }
+
+        [Test]
+        public async Task GetWeatherAsync_Sucess()
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.Tests/Service/WeatherServiceTest.cs]: Read about teat naming approaches

File: Weather.СonsoleApp/Weather.Tests/Service/WeatherServiceTest.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        {
+            var weater = new WeatherResponse()
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.Tests/Service/WeatherServiceTest.cs]: AAA //Arrange //Act //Assert

**Review by Xodoc — APPROVED** [review]

**Review by ihhyy — APPROVED** [review]

File: Weather.СonsoleApp/Weather.Tests/Class1.cs [hunk 3]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Weather.Tests
+{
+    public class Class1
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.Tests/Class1.cs]: ?

File: Weather.СonsoleApp/Weather.Tests/Service/ValidatorTest.cs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+        [SetUp]
+        public void Setup()
+        {
+            _validator = new Validator();
+        }
+
+        [TestCase("")]
+        [TestCase(null)]
+        [Test]
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.Tests/Service/ValidatorTest.cs]: ?

File: Weather.СonsoleApp/Weather.Tests/Service/ValidatorTest.cs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        public void ValidateCityByName_IfStringEmpty_ValidationIsFaild(string name)
+        {
+            // Arrange
+
+            // Act
+
+            // Assert
+            Assert.Throws<ValidationException>(() => _validator.ValidateCityByName(name));
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.Tests/Service/ValidatorTest.cs]: Check error text

File: Weather.СonsoleApp/Weather.Tests/Service/ValidatorTest.cs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        [SetUp]
+        public void Setup()
+        {
+            _validator = new Validator();
+        }
+
+        [TestCase("")]
+        [TestCase(null)]
+        [Test]
+        public void ValidateCityByName_IfStringEmpty_ValidationIsFaild(string name)
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.Tests/Service/ValidatorTest.cs]: Check the name

File: Weather.СonsoleApp/Weather.Tests/Service/ValidatorTest.cs [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        public void ValidateCityByName_IfStringEmpty_ValidationIsFaild()
+        {
+            // Arrange
+            var name = string.Empty;    
+
+            // Act
+            _validator.ValidateCityByName(name);
+
+            // Assert
+            Assert.Throws<ValidationException>(() => _validator.ValidateCityByName(name));
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.Tests/Service/ValidatorTest.cs]: Add success path

File: Weather.СonsoleApp/Weather.Tests/Service/WeatherServiceTest.cs [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+            {
+                Name = "Minsk",
+                Main = new TemperatureInfo() { Temp = 5, }
+
+            };
+            // Act
+            var name = "Minsk";
+            _validatorMock.Verify(x => x.ValidateCityByName(name), Times.Never());
+            _weatherRepositoryMock.Setup(x => x.GetWeatherAsync(name)).ReturnsAsync(weater);
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.Tests/Service/WeatherServiceTest.cs]: arrange

**Review by denja2 — CHANGES_REQUESTED** [review]

File: Weather.СonsoleApp/Weather.Tests/Service/WeatherServiceTest.cs (hunk 1 above)
Comment (alesyalis) [reply Weather.СonsoleApp/Weather.Tests/Service/WeatherServiceTest.cs]: done

File: Weather.СonsoleApp/Weather.Tests/Service/WeatherServiceTest.cs (hunk 2 above)
Comment (alesyalis) [reply Weather.СonsoleApp/Weather.Tests/Service/WeatherServiceTest.cs]: done

File: Weather.СonsoleApp/Weather.Tests/Class1.cs (hunk 3 above)
Comment (alesyalis) [reply Weather.СonsoleApp/Weather.Tests/Class1.cs]: done

File: Weather.СonsoleApp/Weather.Tests/Service/ValidatorTest.cs (hunk 4 above)
Comment (alesyalis) [reply Weather.СonsoleApp/Weather.Tests/Service/ValidatorTest.cs]: done

File: Weather.СonsoleApp/Weather.Tests/Service/ValidatorTest.cs (hunk 5 above)
Comment (alesyalis) [reply Weather.СonsoleApp/Weather.Tests/Service/ValidatorTest.cs]: done

File: Weather.СonsoleApp/Weather.Tests/Service/ValidatorTest.cs (hunk 6 above)
Comment (alesyalis) [reply Weather.СonsoleApp/Weather.Tests/Service/ValidatorTest.cs]: done

File: Weather.СonsoleApp/Weather.Tests/Service/WeatherServiceTest.cs (hunk 8 above)
Comment (alesyalis) [reply Weather.СonsoleApp/Weather.Tests/Service/WeatherServiceTest.cs]: done

File: Weather.СonsoleApp/Weather.BL/DTOs/WeatherResponseNew.cs [hunk 9]
```diff
@@ -0,0 +1,8 @@
+﻿namespace Weather.BL.DTOs
+{
+    public class WeatherResponseNew
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.BL/DTOs/WeatherResponseNew.cs]: Remove new

File: Weather.СonsoleApp/Weather.BL/Services/Abstract/IWeatherService.cs [hunk 10]
```diff
@@ -5,6 +5,7 @@ namespace Weather.BL.Services.Abstract
 {
     public interface IWeatherService
     {
-        Task<WeatherResponseDTO> GetWeatherAsync(string cityName);
+        // Task<string> GetWeatherAsync(string cityName);
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.BL/Services/Abstract/IWeatherService.cs]: remove

File: Weather.СonsoleApp/Weather.Tests/Service/ValidatorTest.cs (hunk 5 above)
Comment (denja2) [reply Weather.СonsoleApp/Weather.Tests/Service/ValidatorTest.cs]: ??

File: Weather.СonsoleApp/Weather.Tests/Service/ValidatorTest.cs [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+        {
+            // Arrange
+            var name = string.Empty;    
+
+            // Act
+            void result () => _validator.ValidateCityByName(name);
+
+            // Assert
+            Assert.Throws<ValidationException>(result);
+        }
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.Tests/Service/ValidatorTest.cs:31]: remove

File: Weather.СonsoleApp/Weather.Tests/Service/ValidatorTest.cs [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+
+        [Test]
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.Tests/Service/ValidatorTest.cs]: TestCase null and empty

File: Weather.СonsoleApp/Weather.Tests/Service/ValidatorTest.cs [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+            var message = "Entering the city is required\n";
+
+            // Act
+            void result() => _validator.ValidateCityByName(name);
+            Exception ex = Assert.Throws<ValidationException>(result);
+
+            // Assert
+            Assert.AreEqual(message, ex.Message);
+        }
+    }
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.Tests/Service/ValidatorTest.cs:44]: add success pass

File: Weather.СonsoleApp/Weather.Tests/Service/WeatherServiceTest.cs [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+        {
+            _weatherRepositoryMock = new Mock<IWeatherRepository>();
+            _validatorMock = new Mock<IValidator>();
+
+            _weatherService = new WeatherService(
+                _weatherRepositoryMock.Object,
+                _validatorMock.Object);
+        }
+
+        [Test]
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.Tests/Service/WeatherServiceTest.cs]: Use testcase

File: Weather.СonsoleApp/Weather.Tests/Service/WeatherServiceTest.cs [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+            // Arrange
+            var weather = new WeatherResponse()
+            {
+                Name = "Minsk",
+                Main = new TemperatureInfo() { Temp = -1, Description = "Dress warmly." }
+            };
+            var name = "Minsk";
+            var message = $"In {weather.Name}: {weather.Main.Temp} °C {weather.Main.Description} ";
+            
+            _validatorMock.Verify(x => x.ValidateCityByName(name), Times.Never());
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.Tests/Service/WeatherServiceTest.cs]: Move to Assert

File: Weather.СonsoleApp/Weather.Tests/Service/WeatherServiceTest.cs [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+            _weatherRepositoryMock.Setup(x => x.GetWeatherAsync(name)).ReturnsAsync(weather);
+            
+            // Act
+            var result = await _weatherService.GetWeatherAsync(name);
+            
+            // Assert
+            Assert.AreEqual(message, result.Message);
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.Tests/Service/WeatherServiceTest.cs]: is error

File: Weather.СonsoleApp/Weather.Tests/Service/WeatherServiceTest.cs [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+                Main = new TemperatureInfo() { Temp = 35, Description = "It's time to go to the beach" }
+            };
+            var name = "Minsk";
+            _weatherRepositoryMock.Setup(x => x.GetWeatherAsync(name)).ReturnsAsync(() => weather);
+
+            // Act
+            var result = await _weatherService.GetWeatherAsync(name);
+
+            // Assert
+            Assert.IsFalse(result.IsError);
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.Tests/Service/WeatherServiceTest.cs]: If exception is thrown from repo and validator

File: Weather.СonsoleApp/Weather.СonsoleApp/Program.cs [hunk 18]
```diff
@@ -30,14 +28,10 @@ static async Task Main(string[] args)
...
-                catch (ArgumentNullException ex)
-                {
-                    Console.WriteLine(ex.Message);
+                    Console.WriteLine("Enter the name of the city");
+                    var cityName = Console.ReadLine();
+                    var weather = await _weatherService.GetWeatherAsync(cityName);
+                    Console.WriteLine(weather.Message);
                 }
                 catch (ValidationException ex)
                 {
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.СonsoleApp/Program.cs]: Unhandled error

File: Weather.СonsoleApp/Weather.Tests/Service/ValidatorTest.cs (hunk 7 above)
Comment (alesyalis) [reply Weather.СonsoleApp/Weather.Tests/Service/ValidatorTest.cs]: done

File: Weather.СonsoleApp/Weather.Tests/Service/WeatherServiceTest.cs [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+            var result = await _weatherService.GetWeatherAsync(name);
+
+            // Assert
+            _validatorMock.Verify(x => x.ValidateCityByName(name), Times.Once());
+            Assert.IsNotNull(result);
+            Assert.AreEqual(message, result.Message);
+            Assert.IsTrue(result.IsError);
+        }
+
+        [TestCase("")]
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.Tests/Service/WeatherServiceTest.cs]: Test

File: Weather.СonsoleApp/Weather.Tests/Service/WeatherServiceTest.cs [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+        {
+            // Arrange
+            var expectedExcetpion = new Exception();
+            _weatherRepositoryMock.Setup(x => x.GetWeatherAsync(name)).ReturnsAsync(() => throw  new Exception());
+
+            // Act
+            // Assert
+            var result = Assert.ThrowsAsync<Exception>( () => _weatherService.GetWeatherAsync(name));
+            Assert.AreEqual(expectedExcetpion.Message, result.Message);  
+        }
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.Tests/Service/WeatherServiceTest.cs:87]: Case with exception from validator

File: Weather.СonsoleApp/Weather.IntegrationTest/IntegrationTest.cs [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Text;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Weather.IntegrationTest
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.IntegrationTest/IntegrationTest.cs]: Remove integration tests from pr

**Review by denja2 — APPROVED** [review]

**Pull request merged**